lang/cs/Org.Apache.REEF.Network/Elastic/Task/CommunicationLayer.cs

```diff
@@ -0,0 +1,283 @@
+            }
+        }
+
+        /// <summary>
+        /// Registers a <see cref="DriverAwareOperatorTopology"/> with the communication layer.
+        /// </summary>
+        /// <param name="operatorObserver">The observer of the driver aware topology</param>
+        internal void RegisterOperatorTopologyForDriver(DriverAwareOperatorTopology operatorObserver)
+        {
+            if (!_driverMessageObservers.TryAdd(operatorObserver.NodeId(), operatorObserver))
+            {
+                throw new IllegalStateException($"Topology for id {operatorObserver.NodeId()} already added among driver listeners.");
+            }
+        }
+
+        /// <summary>
+        /// Send the communication message to the task whose name is included in the message.
+        /// </summary>
+        /// <param name="destination">The destination node for the message</param>
+        /// <param name="message">The message to send</param>
+        /// <param name="cancellationSource">The token to cancel the operation</param>
+        internal void Send(
+            string destination,
+            ElasticGroupCommunicationMessage message,
+            CancellationTokenSource cancellationSource)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException(nameof(message));
+            }
+            if (string.IsNullOrEmpty(destination))
+            {
+                throw new ArgumentNullException(nameof(destination));
+            }
+            if (_disposed)
+            {
+                Log.Log(Level.Warning, "Received send message request after disposing: Ignoring.");
+                return;
+            }
+
+            IIdentifier destId = _idFactory.Create(destination);
+
+            for (int retry = 0;  !Send(destId, message); retry++)
+            {
+                if (retry > _retrySending)
+                {
+                    throw new IllegalStateException($"Unable to send message after retrying {retry} times.");
+                }
+                Thread.Sleep(_timeout);
+            }
+        }
+
+        /// <summary>
+        /// Forward the received message to the target <see cref="IOperatorTopologyWithCommunication"/>.
+        /// </summary>
+        /// <param name="remoteMessage">The received message</param>
+        public abstract void OnNext(IRemoteMessage<NsMessage<ElasticGroupCommunicationMessage>> remoteMessage);
+
+        /// <summary>
+        /// Checks if the identifier is registered with the name server.
+        /// Throws exception if the operation fails more than the retry count.
+        /// </summary>
+        /// <param name="identifiers">The identifier to look up</param>
+        /// <param name="cancellationSource">The token to cancel the operation</param>
+        /// <param name="removed">Nodes that got removed during task registration</param>
+        public void WaitForTaskRegistration(
+            ICollection<string> identifiers,
```

Comment:
better, `IEnumerable<string> identifiers` - then we don't have to convert, e.g., dictionary values with `.ToList()` explicitly, etc.

Reply:
Fixed.